### PR TITLE
feat(substrait): modular substrait consumer

### PR DESCRIPTION
**Pull request by vbarua**
Note: When reviewing this PR I would recommend hiding whitespace changes.

## Which issue does this PR close?
Initial work for https://github.com/apache/datafusion/issues/13318

## Rationale for this change
Improves the reusability of the Substrait consumer for users that utilize user-defined extensions and types.

I should note that this PR represents an approach to the problem, but is not necessarily the only way we could handle it. I'm very open to feedback on the approach.

## What changes are included in this PR?
* Relation and expression handling code has been extracted into a series of `from_substrait_*` functions (i.e. `from_filter_rel`, `from_scalar_function`) to aid re-use.
* A `SubstraitConsumer` trait has been introduced with default methods that handle relations and expression using the above functions.
* The `SubstraitConsumer` trait also includes methods for handling extension relations (i.e `consume_extension_single`) along with user-defined types (`consume_user_defined_type`, `consume_user_defined_literal`).
* All conversion methods take as their first argument an `&impl SubstraitConsumer`.
* The role of `SubstraitPlanningState` has been subsumed into the `SubstraitConsumer`.

## Are these changes tested?
These changes refactor existing code and leverage their tests.

A doc comment has been added to show how the `SubstraitConsumer` trait can be implemented by users.

## Are there any user-facing changes?
Yes.

The top-level `from_substrait_plan(state, plan)` function now takes a `&SessionState` as its first argument, instead of `&dyn SubstraitPlanningState`. The same is true for `from_substrait_extended_expr`. These two functions are the primary external facing APIs for intiation a conversion.

A number of functions like `from_substrait_rel`, `from_substrait_rex`, etc have had their API changed to make the first argument `&impl SubstraitConsumer`, and remove all `state: &dyn SubstraitPlanningState` and `extensions: &Extensions` arguments.

### CODE REVIEW

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:1081]: These Extension relation handling code has been fully removed as that must be provided by users.

**Comment by vbarua** [comment]
@Blizzara I would appreciate if you could take a look and let me know if this approach seems reasonable, or if I've [architecture astronauted](https://en.wikipedia.org/wiki/Architecture_astronaut) myself.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:1056]: The usage of `SerializerRegistry` is problematic here because `deserialize_logical_plan` assumes that the extension details are a fully formed `UserDefinedLogicalNode`. This _may_ sometimes be the case, but the purpose of the Substrait Extension relations is to allow users to encoded arbitrary relations into Substrait in a manner they see fit.

The only really foolproof way of handling this is to provide a hook like `consume_extension_single` which users can  use to provide their deserialisation logic.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:1424]: Hm, these aren't used anywhere, and aren't they kinda weirdly opposite of what the other functions do (ie these call consumer.consume.., instead of consumer.consume.. calling these), maybe let's not add them?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:1081]: Would it make sense to keep these in the default implementation though? Just as in the producer we do produce these types (https://github.com/apache/datafusion/blob/b8e435f3e38f1f8d2d8f24c3da3c2b21e76b700b/datafusion/substrait/src/logical_plan/producer.rs#L596), and I guess it'd be nice if the roundtrip tests that don't require special setup on producer wouldn't require special setup for consumer either.

That's not a strictly held opinion.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:1869]: You mentioned you'd prefer to not have the get_extensions(), I guess you could replace it with something like `get_function` which in DefaultSubstraitConsumer would just do this `extension.functions.get(..)`? Or did you have something else in mind?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:1961]: should this be renamed to `from_field_reference`? I saw this is used for RexType::Selection, but maybe it'd still make more sense to name it after the parameter type..?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:1081]: > Would it make sense to keep these in the default implementation though?

I think it would be reasonable to include these in the `DefaultSubstraitConsumer`. That would play nice with roundtrip tests, and it might be a reasonable default to assume that extension relations are `UserDefinedLogicalNode`s.

I'll add these back in.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:2445]: I think the producer doesn't produce any of these anymore, we just didn't remove them yet in case someone is producing plans with an older version and trying to run them with a newer one. Or like, run a plan that has been stored somewhere. But maybe it's time to remove them from the next version for example.

**Comment by Blizzara** [comment]
> @Blizzara I would appreciate if you could take a look and let me know if this approach seems reasonable, or if I've [architecture astronauted](https://en.wikipedia.org/wiki/Architecture_astronaut) myself.

Thanks for the ping - I took a look, and overall looks really good to me! I left couple thoughts (I see you already saw some), but nothing major. I already have usecase in mind for this internally, so looking forward to getting it merged 😄 

One thing is we should consider splitting the consumer.rs into smaller files. It could be a separate PR to keep the diff smaller, but maybe moving all the static "consumer_X" functions into like `consumer/relations.rs`, `consumer/expressions.rs`, `consumer/literals.rs` and `consumer/types.rs` (and one more for utils etc) or similar? Feels like now could be a good moment to do that.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs
Comment (ccciudatu) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:1019]: One drawback of the current implementation is that it's abusing the `type_url` field in `proto::Any` (which is quite [clearly specified](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/AnyOrBuilder.html#getTypeUrl--). Working with Substrait objects directly would eliminate the need for hacks like encoding node names as Any::type_url.

As a side note, this may sound hypocritical, given that I made it even worse [with my PR](https://github.com/apache/datafusion/pull/13772) by setting it to table names as well. But I only did it for consistency. :)

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: These were leftover from an earlier approach I took to this and aren't needed anymore. I've removed them.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: I think that's a better name. I updated both `consume_selection` and `from_selection`.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:1056]: I added the SerializerRegistry based code back into the DefaultSubstraitConsumer for now.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:2483]: I thought about removing it as part of this PR, but figured that would be better handled as a separate follow-up change. I'm already breaking enough API with this as is 😅

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:1907]: I was trying to figure out the right abstraction not just for resolving the function via the extension, but also mapping it into DataFusion. However I found the existing logic a little unwieldy to reason about. Specifically scalar functions first try and lookup the function by name, then look it up as an operator and then try and treat is a BuiltinExpr.

I want to think about function mapping in isolation, because I suspect it will be easier to figure out the right abstraction if we simplify the lookup logic first.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs
Comment (vbarua) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs]: > As a side note, this may sound hypocritical, given that I made it even worse https://github.com/apache/datafusion/pull/13772

I don't think it's hypocritical, you were aiming for consistency and I'm just being much more aggressive with my changes and breaking the API.

**Comment by vbarua** [comment]
@Blizzara 
> One thing is we should consider splitting the consumer.rs into smaller files. It could be a separate PR to keep the diff smaller, but maybe moving all the static "consumer_X" functions into like consumer/relations.rs, consumer/expressions.rs, consumer/literals.rs and consumer/types.rs (and one more for utils etc) or similar? Feels like now could be a good moment to do that.

I think it makes sense to do that, and I'm happy to own it, but I would prefer to do that as a followup. I structured the `consumer.rs` changes in order to make it easy to see that the functions I added were extracted from the big `from_substrait_rel` and `from_substrait_rex` switch statements.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:1907]: > scalar functions first try and lookup the function by name, then look it up as an operator and then try and treat is a BuiltinExpr.

Yeah, I think this is a side-effect of DF not having (yet?) converged fully on UDFs. I think for aggregates and windows the built-in modes were already removed in favor of UDAF/UDWF but not for scalars.

But yep, totally good to do as a followup someday!

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:228]: It looks like `get_state()` is only used for getting udf/udaf/udwfs? So if you'd wish, you could replace SessionState with FunctionRegistry. Then the change would be compatible with anyone using the current `SubstraitPlanningState` rather than a SessionState, since SubstraitPlanningState requires FunctionRegistry.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:137]: do you need this? SubstraitPlanningState doesn't seem to be used anywhere in the code anymore afaict. Actually, if it's not used anymore after this, maybe remove the whole state.rs file as part of this PR?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:41]: here as well, is this actually used? I don't see any references to it anymore

**Comment by Blizzara** [comment]
> I think it makes sense to do that, and I'm happy to own it, but I would prefer to do that as a followup. I structured the `consumer.rs` changes in order to make it easy to see that the functions I added were extracted from the big `from_substrait_rel` and `from_substrait_rex` switch statements.

Yup, I totally agree!

I looked through the latest changes, looks even better now, just some more comments re SubstraitPlanningState but otherwise this looks very good by me!

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:41]: `SubstraitPlanningState` is used in `producer.rs` still, so we can't remove it yet.

The import here brings the `table()` fn here https://github.com/apache/datafusion/blob/82a40f3edf264834c28f3a1bf2a3143951802634/datafusion/substrait/src/logical_plan/state.rs#L54-L62
into scope. It could make sense to inline it to remove it.

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:41]: Ah I see, yep I'd be for inlining it here, but not a critical thing then!

Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Went ahead and inlined it to be able to remove all usages from the consumer. Will make it easier to remove fully once we updated the producer as well.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Removed

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Good idea, it's a more precise constraint to have.

**Review by Blizzara — APPROVED** [review]
LGTM 🚀 Thanks @vbarua, and FYI @alamb :)

**Review by alamb — COMMENTED** [review]
Thank you @vbarua and @Blizzara  -- this looks like a great improvement to me

I found another bug while reviewing this PR 🤦 
- https://github.com/apache/datafusion/issues/13854

The only thing I think we should try and fix before merging this is needing to copy the session context into an Arc

I realize the issue is related to fighting the borrow checker -- I played around with it and came up with this which seemed to work:
- https://github.com/vbarua/datafusion/pull/1


While it will be potentially disruptive to anyone using the Substrait bindings I think this is necessary to get the APIs in shape to support user defined functionaly as you point out. 

I really like the fact that the existing high level API to create subtrait doesn't change

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:223]: Created https://github.com/apache/datafusion/issues/13863 for this linked it in the code.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: In the docs, I've pointed users towards `from_substrait_plan` which uses this.

This is really only `pub` so it can be used in integration tests.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs
Comment (vbarua) [reply datafusion/substrait/tests/cases/roundtrip_logical_plan.rs]: I've updated this to call `from_substrait_plan`. This is a holdover from an earlier design in which the DefaultSubstraitConsumer didn't handle extension relations

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Replaced with lifetime annotations based on @alamb proposals in https://github.com/vbarua/datafusion/pull/1

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:231]: @/Blizzara also suggested splitting the code into modules. I've combined both of those ideas into a future issue https://github.com/apache/datafusion/issues/13864

**Review by alamb — APPROVED** [review]
Love it -- thank you @vbarua and thank you @Blizzara for your comments and feedback

**Comment by alamb** [comment]
I'll plan to merge this tomorrow unless anyone would like some more time to review

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:486]: In a prior versions this used to implement `Default`. I've removed it as it was intended mostly for testing. External users should just call `from_substrait_plan` as before.

Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:486]: In a prior versions this used to implement `Default`. I've removed it as it was intended mostly for testing. External users should just call `from_substrait_plan` as before.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:739]: nit: I think if you want to, you could remove the Extension lines from below - no need to duplicate them there

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:731]: I started thinking whether it actually  really make sense to pass the Extensions into the SubstraitConsumer, or if there should be some `initialize(plan: &Plan)` method or something in the SubstraitConsumer that loads them. But I didn't think yet much about it, maybe there is a reason to rather have it like this, and anyways no need to change anything for this PR anymore in any case 😄

Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:731]: I do want to streamline the API further, but in the future 😅

Initialising the consumer directly from the plan makes sense to me, because we always need the extensions.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:739]: That is redundant now isn't it. Removed.

**Review by alamb — COMMENTED** [review]
👍

**Comment by alamb** [comment]
FYI @ccciudatu

**Pull request merged**
